xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/internal/DefaultIndexingUserConfig.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.search.solr.internal;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.search.solr.internal.api.IndexingUserConfig;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.classes.ListClass;
+
+/**
+ * Provides the configured indexing user via information stored in a wiki page.
+ *
+ * @version $Id$
+ * @since 14.8
+ */
+@Component
+@Singleton
+public class DefaultIndexingUserConfig implements IndexingUserConfig
+{
+    /**
+     * The class storing the indexing user.
+     */
+    static final LocalDocumentReference CONFIG_CLASS
+        = new LocalDocumentReference(XWiki.SYSTEM_SPACE, "SolrSearchAdminIndexingUserClass");
+
+    /**
+     * The wiki page storing the indexing user.
+     */
+    static final LocalDocumentReference CONFIG_PAGE
+        = new LocalDocumentReference(XWiki.SYSTEM_SPACE, "SolrSearchAdminIndexingUser");
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> documentReferenceResolver;
+
+    @Override
+    public DocumentReference getIndexingUserReference() throws XWikiException
+    {
+        DocumentReference indexingUser = null;
+        XWikiContext context = contextProvider.get();
+        XWiki wiki = context.getWiki();
+
+        XWikiDocument configPage = wiki.getDocument(new DocumentReference(CONFIG_PAGE,
+            context.getWikiReference()), context);
```

Comment:
You said in the description that this is supposed to be configured on the main wiki but here you are loading the configuration in the current wiki (you could have directly passed the `LocalDocumentReference` to `getDocument` by the way, no need to create a `DocumentReference`).

Reply:
Ah, I did not know that I can use a `LocalDocument` reference.
The current wiki where the reference is loaded is always the main wiki - that is not decided here, however, but how the `DefaultSolrIndexer` sets up it the ExecutionContext.  
If that setup ever changes then indeed the statement about the main wiki is wrong, but I am happy to update the code to support different indexing users in subwikis then.

Reply:
>  The current wiki where the reference is loaded is always the main wiki - that is not decided here, however, but how the DefaultSolrIndexer sets up it the ExecutionContext.

We are not talking about a private method in `DefaultSolrIndexer` here, but an independent component which might be used elsewhere some day without thinking what it does is actually wrong. If the configuration only make sense at main wiki level, `DefaultIndexingUserConfig` should reflect that.

Reply:
Well, I hoped in the future to support also use cases where the indexing user can be defined separately in a subwiki (it might happen that non-admin users in the main wiki have no access to the subwikis). This seems currently not possible with the way the indexer works (unless the "batches" always belong to one subwiki, or the user is set in a different place).

But anyway, as the case is now, I bulletproofed the code to always fetch the indexing user from the main wiki.